File: csharp/DVSClient/DVSClient/email/validate/ValidateResultFuture.cs
using DVSClient.Exceptions;
using DVSClient.Server.Email;

namespace DVSClient.Email.Validate
{
    public class ValidateResultFuture
    {
        private readonly Task<RestApiEmailValidateResponse> _apiFuture;

        public ValidateResultFuture(Task<RestApiEmailValidateResponse> apiFuture)
        {
            _apiFuture = apiFuture;
        }

        public async Task<ValidateResult> GetAsync()
        {
            try
            {
                var response = await _apiFuture;

                if (response.Error != null)
                {
                    throw EDVSException.Using(response.Error);
                }
                return new ValidateResult(response);
            }
            catch (Exception e) when (e is TaskCanceledException || e is OperationCanceledException)
            {
                throw new RestApiInterruptionOrExecutionException(e);
            }
        }

        public async Task<ValidateResult> GetAsync(TimeSpan timeout)
        {
            try
            {
                var response = await Task.WhenAny(_apiFuture, Task.Delay(timeout)) == _apiFuture
                    ? await _apiFuture
                    : throw new TimeoutException();

                if (response.Error != null)
                {
                    throw EDVSException.Using(response.Error);
                }
                return new ValidateResult(response);
            }
            catch (Exception e) when (e is TaskCanceledException || e is OperationCanceledException || e is TimeoutException)
            {
                throw new RestApiInterruptionOrExecutionException(e);
            }
        }

        public bool Cancel(bool mayInterruptIfRunning)
        {
            return _apiFuture.IsCanceled;
        }

        public bool IsCancelled()
        {
            return _apiFuture.IsCanceled;
        }

        public bool IsDone()
        {
            return _apiFuture.IsCompleted;
        }
    }
}

File: java/src/main/java/com/experian/dvs/client/email/validate/ValidateResultFuture.java
package com.experian.dvs.client.email.validate;

import com.experian.dvs.client.exceptions.EDVSException;
import com.experian.dvs.client.server.email.RestApiEmailValidateResponse;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class ValidateResultFuture implements Future<ValidateResult> {

    private final Future<RestApiEmailValidateResponse> apiFuture;

    public ValidateResultFuture(Future<RestApiEmailValidateResponse> apiFuture) {
        this.apiFuture = apiFuture;
    }

    @Override
    public boolean cancel(boolean mayInterruptIfRunning) {
        return this.apiFuture.cancel(mayInterruptIfRunning);
    }

    @Override
    public boolean isCancelled() {
        return this.apiFuture.isCancelled();
    }

    @Override
    public boolean isDone() {
        return this.apiFuture.isDone();
    }

    @Override
    public ValidateResult get() throws InterruptedException, ExecutionException {
        final RestApiEmailValidateResponse response = this.apiFuture.get();
        if (response.getError() != null) {
            throw EDVSException.using(response.getError());
        }
        return new ValidateResult(response);
    }

    @Override
    public ValidateResult get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
        final RestApiEmailValidateResponse response = this.apiFuture.get(timeout, unit);
        if (response.getError() != null) {
            throw EDVSException.using(response.getError());
        }
        return new ValidateResult(response);
    }
}
